### ISSUE TITLE
why is the magnitude of input_image not divided?

### ISSUE THREAD

**Issue by yoon28** [AntreasAntoniou/MatchingNetworks#15]
To compute cosine similarity, the dot product of two vectors is divided by the magnitudes of both vectors. 
i.e. cos(A,B) = dot(A,B)/(||A||*||B||).
However, you did not use the magnitude of input_image to compute cosine similarity.
Why is that?

```
class DistanceNetwork():

    def __init__(self):
        self.reuse = False

    def __call__(self, support_set, input_image, name, training=False):
        with tf.name_scope('distance-module' + name), tf.variable_scope('distance-module', reuse=self.reuse):
            eps = 1e-10
            similarities = []
            for support_image in tf.unstack(support_set, axis=0):
                sum_support = tf.reduce_sum(tf.square(support_image), 1, keep_dims=True)
                support_magnitude = tf.rsqrt(tf.clip_by_value(sum_support, eps, float("inf")))
                dot_product = tf.matmul(tf.expand_dims(input_image, 1), tf.expand_dims(support_image, 2))
                dot_product = tf.squeeze(dot_product, [1, ])
                cosine_similarity = dot_product * support_magnitude
                similarities.append(cosine_similarity)

        similarities = tf.concat(axis=1, values=similarities)
        self.variables = tf.get_collection(tf.GraphKeys.TRAINABLE_VARIABLES, scope='distance-module')

        return similarities
```

**Comment by AntreasAntoniou** [comment]
Initially I went that way and with the setup I had then the optimization
was more unstable, so I had a look in other optimizations and found that
others did not scale by the target as well. Sometimes having too small a
loss can cause gradient propagation issues. However since then I improved
multiple parts of the system so using both support and target magnitudes
might not be such a bad idea. Have a try if you want and let me know if it
works.
